src/Common/SipHash.h

```diff
@@ -29,6 +31,61 @@ namespace ErrorCodes
 {
     extern const int LOGICAL_ERROR;
 }
+
+template <typename T>
+struct TransformToLittleEndian
+{
+    T operator()(const T & x) const noexcept
+    {
+        static_assert(std::is_scalar_v<T>);
+
+        auto rev_x = x;
+        auto * start = reinterpret_cast<std::byte *>(&rev_x);
+        auto * end = start + sizeof(T);
+        std::reverse(start, end);
+
+        return rev_x;
+    }
+};
+
+template <size_t Bits, typename Signed>
+struct TransformToLittleEndian<wide::integer<Bits, Signed>>
+{
+    wide::integer<Bits, Signed> operator()(const wide::integer<Bits, Signed> & x) const noexcept
+    {
+        auto transformed = x;
+        std::ranges::reverse(transformed.items);
+        std::ranges::transform(transformed.items, std::begin(transformed.items), std::byteswap<typename wide::integer<Bits, Signed>::base_type>);
+        return transformed;
+    }
+};
+
+template <>
+struct TransformToLittleEndian<IPv6>
+{
+    IPv6 operator()(const IPv6 & x) const noexcept
+    {
+        return IPv6(TransformToLittleEndian<IPv6::UnderlyingType>()(x.toUnderType()));
+    }
+};
+
+template <typename T, typename Tag>
+struct TransformToLittleEndian<StrongTypedef<T, Tag>>
+{
+    StrongTypedef<T, Tag> operator()(const StrongTypedef<T, Tag> & x) const noexcept
+    {
+        return StrongTypedef<T, Tag>(TransformToLittleEndian<typename StrongTypedef<T, Tag>::UnderlyingType>()(x.toUnderType()));
+    }
+};
+
+template <typename A, typename B>
+struct TransformToLittleEndian<std::pair<A, B>>
+{
+    std::pair<A, B> operator()(const std::pair<A, B> & x) const noexcept
+    {
+        return {TransformToLittleEndian<A>()(x.first), TransformToLittleEndian<B>()(x.second)};
```

Comment:
You got everything perfectly right. I kind of wished to err on the side of backward compatibility, since these hashes are used in all sorts of places starting from network communication to file checksums. The textbook solution would be to reverse the "outer" order as well, everywhere, but that'd require much more tinkering. With that being said, if you're willing to go down the route of modifying CityHash I could evaluate how much effort it'd all entail. What do you think?

Reply:
My intention with mentioning CityHash is to state where the confusion is coming from. While it does handle endianness in terms of the individual components (https://github.com/ClickHouse/ClickHouse/blob/master/contrib/cityhash102/src/city.cc) it seems to explicitly define ordering between the components themselves (https://github.com/ClickHouse/ClickHouse/blob/master/contrib/cityhash102/include/city.h) with `Uint128Low64` and `Uint128High64`. Were these following a textbook approach, they'd be returning the opposite component on big-endian. If we were to change these, it may or may not end up working in the end and would require us to flip the individual components as well upon serialization / deserialization.

For now, I suggest that we don't tamper with CityHash and thus keep the status quo even if for just a little longer so as not to grow this PR to an unmanageable size.

For the future, would you be willing to accept changes to CityHash itself in this regard in order to make it right?

Reply:
> For the future, would you be willing to accept changes to CityHash itself in this regard in order to make it right?

Yepp. Also agree that we should not blow this PR further up.